File: felicia/examples/learn/message_communication/protobuf/python/simple_publishing_node.py
import felicia_py as fel
from felicia.core.protobuf.channel_pb2 import ChannelDef

from felicia.examples.learn.message_communication.protobuf.message_spec_pb2 import MessageSpec


class SimplePublishingNode(fel.NodeLifecycle):
    def __init__(self, topic, channel_type):
        super().__init__()
        self.topic = topic
        self.channel_def = ChannelDef()
        self.channel_def.type = ChannelDef.Type.Value(channel_type)
        self.publisher = fel.Publisher()
        self.message_id = 0

    def on_init(self):
        print("SimplePublishingNode.on_init()")

    def on_did_create(self, node_info):
        print("SimplePublishingNode.on_did_create()")
        self.node_info = node_info
        self.request_publish()

        # fel.MasterProxy.post_delayed_task(
        #     self.request_unpublish, fel.TimeDelta.from_seconds(10))

    def on_error(self, status):
        print("SimplePublishingNode.on_error()")
        fel.log(fel.ERROR, status.error_message())

    def request_publish(self):
        settings = fel.Settings()
        settings.buffer_size = fel.Bytes.from_bytes(512)

        self.publisher.request_publish(
            self.node_info, self.topic, self.channel_def, settings,
            self.on_request_publish)

    def on_request_publish(self, status):
        print("SimplePublishingNode.on_request_publish()")
        fel.log_if(fel.ERROR, not status.ok(), status.error_message())
        self.repeating_publish()

    def repeating_publish(self):
        self.publisher.publish(self.generate_message(), self.on_publish)

        if not self.publisher.is_unregistered():
            fel.MasterProxy.post_delayed_task(
                self.repeating_publish, fel.TimeDelta.from_seconds(1))

    def on_publish(self, status):
        print("SimplePublishingNode.on_publish()")
        fel.log_if(fel.ERROR, not status.ok(), status.error_message())

    def generate_message(self):
        message_spec = MessageSpec()
        message_spec.id = self.message_id
        message_spec.timestamp = fel.Time.now().to_double_t()
        message_spec.content = "hello world"
        self.message_id += 1
        return message_spec

    def request_unpublish(self):
        self.publisher.request_unpublish(self.node_info, self.topic,
                                         self.on_request_unpublish)

    def on_request_unpublish(self, status):
        print("SimplePublishingNode.on_request_unpublish()")
        fel.log_if(fel.ERROR, not status.ok(), status.error_message())

File: felicia/examples/learn/message_communication/protobuf/cc/simple_publishing_node.cc
#include "felicia/examples/learn/message_communication/protobuf/cc/simple_publishing_node.h"

#include "felicia/core/master/master_proxy.h"

namespace felicia {

SimplePublishingNode::SimplePublishingNode(
    const NodeCreateFlag& node_create_flag,
    SSLServerContext* ssl_server_context)
    : node_create_flag_(node_create_flag),
      topic_(node_create_flag_.topic_flag()->value()),
      ssl_server_context_(ssl_server_context) {}

void SimplePublishingNode::OnInit() {
  std::cout << "SimplePublishingNode::OnInit()" << std::endl;
}

void SimplePublishingNode::OnDidCreate(const NodeInfo& node_info) {
  std::cout << "SimplePublishingNode::OnDidCreate()" << std::endl;
  node_info_ = node_info;
  RequestPublish();

  // MasterProxy& master_proxy = MasterProxy::GetInstance();
  // master_proxy.PostDelayedTask(
  //     FROM_HERE,
  //     base::BindOnce(&SimplePublishingNode::RequestUnpublish,
  //                      base::Unretained(this)),
  //     base::TimeDelta::FromSeconds(10));
}

void SimplePublishingNode::OnError(const Status& s) {
  std::cout << "SimplePublishingNode::OnError()" << std::endl;
  LOG(ERROR) << s;
}

void SimplePublishingNode::OnPublish(ChannelDef::Type type, const Status& s) {
  std::cout << "SimplePublishingNode::OnPublish() from "
            << ChannelDef::Type_Name(type) << std::endl;
  LOG_IF(ERROR, !s.ok()) << s;
}

void SimplePublishingNode::OnRequestPublish(const Status& s) {
  std::cout << "SimplePublishingNode::OnRequestPublish()" << std::endl;
  LOG_IF(ERROR, !s.ok()) << s;
  RepeatingPublish();
}

void SimplePublishingNode::OnRequestUnpublish(const Status& s) {
  std::cout << "SimplePublishingNode::OnRequestUnpublish()" << std::endl;
  LOG_IF(ERROR, !s.ok()) << s;
}

void SimplePublishingNode::RequestPublish() {
  communication::Settings settings;
  settings.buffer_size = Bytes::FromBytes(512);
  if (ssl_server_context_) {
    settings.channel_settings.tcp_settings.use_ssl = true;
    settings.channel_settings.tcp_settings.ssl_server_context =
        ssl_server_context_;
  }

  ChannelDef::Type channel_type;
  ChannelDef::Type_Parse(node_create_flag_.channel_type_flag()->value(),
                         &channel_type);

  publisher_.RequestPublish(
      node_info_, topic_, channel_type, settings,
      base::BindOnce(&SimplePublishingNode::OnRequestPublish,
                     base::Unretained(this)));
}

void SimplePublishingNode::RequestUnpublish() {
  publisher_.RequestUnpublish(
      node_info_, topic_,
      base::BindOnce(&SimplePublishingNode::OnRequestUnpublish,
                     base::Unretained(this)));
}

void SimplePublishingNode::RepeatingPublish() {
  publisher_.Publish(GenerateMessage(),
                     base::BindRepeating(&SimplePublishingNode::OnPublish,
                                         base::Unretained(this)));

  if (!publisher_.IsUnregistered()) {
    MasterProxy& master_proxy = MasterProxy::GetInstance();
    master_proxy.PostDelayedTask(
        FROM_HERE,
        base::BindOnce(&SimplePublishingNode::RepeatingPublish,
                       base::Unretained(this)),
        base::TimeDelta::FromSeconds(1));
  }
}

MessageSpec SimplePublishingNode::GenerateMessage() {
  static int id = 0;
  base::TimeDelta timestamp = timestamper_.timestamp();
  MessageSpec message_spec;
  message_spec.set_id(id++);
  message_spec.set_timestamp(timestamp.InMicroseconds());
  message_spec.set_content("hello world");
  return message_spec;
}

}  // namespace felicia